zap/src/main/java/org/zaproxy/zap/extension/ascan/filters/impl/URLPatternScanFilter.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2019 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.ascan.filters.impl;
+
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.regex.Pattern;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.model.HistoryReference;
+import org.zaproxy.zap.extension.ascan.filters.FilterResult;
+import org.zaproxy.zap.extension.ascan.filters.GenericFilterUtility;
+import org.zaproxy.zap.extension.ascan.filters.ScanFilter;
+import org.zaproxy.zap.extension.ascan.filters.UrlPatternFilterData;
+import org.zaproxy.zap.model.StructuralNode;
+
+/**
+ * ScanFilter implementation for filtering based on Url Patterns.
+ *
+ * @author KSASAN preetkaran20@gmail.com
+ * @since 2.9.0
+ */
+public class URLPatternScanFilter implements ScanFilter {
+
+    private static final String FILTER_TYPE = "scan.filter.filterType.URLPattern";
+    private Set<UrlPatternFilterData> urlPatternFilterDataSet = new LinkedHashSet<>();
+
+    public Set<UrlPatternFilterData> getUrlPatternFilterDataSet() {
+        return urlPatternFilterDataSet;
+    }
+
+    public void setUrlPatternFilterDataSet(Set<UrlPatternFilterData> urlPatternFilterDataSet) {
+        this.urlPatternFilterDataSet = urlPatternFilterDataSet;
+    }
+
+    @Override
+    public FilterResult isFiltered(StructuralNode node) {
+        HistoryReference hRef = node.getHistoryReference();
+        if (hRef == null) {
+            return FilterResult.NOT_FILTERED;
+        }
+
+        for (UrlPatternFilterData urlPatternFilterData : this.urlPatternFilterDataSet) {
+            switch (urlPatternFilterData.getFilterCriteria()) {
+                case INCLUDE:
+                    if (urlPatternFilterData.getUrlPatterns().isEmpty()) {
+                        return FilterResult.NOT_FILTERED;
+                    }
+                    for (Pattern pattern : urlPatternFilterData.getUrlPatterns()) {
+                        if (pattern.matcher(hRef.getURI().toString()).matches()) {
```

Comment:
We should allow to override how the test is done. I'd suggest changing `GenericFilterData<T>` to `GenericFilterData<T, V>` and add `boolean matches(V value)` (which would default to, e.g. `values.stream().anyMatch(e -> predicate.test(e, value))`), then `UrlPatternFilterData` can set/use a custom predicate `(x, y) -> x.matcher(y).matches()`. Once that's done this can just extend `AbstractScanFilter` like the other filters do.

Reply:
hmm... this was a very good suggestion. thanks a lot. i have incorporated this too.